### PR TITLE
Avoid stealing focus from the active output window pane

### PR DESCRIPTION
**Pull request by davkean**
Avoid stealing focus when creating our output window pane.

Fixes part of #12202.

~NOTE: I've spent three hours trying to test this but failed miserably due to a number of issues. I'm not going to spend anymore time on it but others are welcome to take my changes, test and merge.~ I've since tested this and it works.

### CODE REVIEW

**Comment by sharwell** [comment]
Is focus only set to the IntelliSense Errors window the first time a message is written to it (i.e. it never occurs again unless the IDE is restarted)?

File: src/VisualStudio/Core/Def/Implementation/Workspace/WorkspaceFailureOutputPane.cs [hunk 1]
```diff
@@ -49,26 +49,56 @@ private IVsOutputWindowPane OutputPane
...
+                    // Output Window panes have two states; initialized and active. The former is used to indicate that the pane
+                    // can be made active ("selected") by the user, the latter indicates that the pane is currently active.
+                    // There's no way to only initialize a pane without also making it active so we remember the last active pane
+                    // and reactivate it after we've created ourselves to avoid stealing focus away from it.
+                    var lastActivePane = GetActivePane(outputWindow);
+
+                    _doNotAccessDirectlyOutputPane = CreateOutputPane(outputWindow);
+
+                    if (lastActivePane != Guid.Empty)
+                        ActivatePane(outputWindow, lastActivePane);
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/Implementation/Workspace/WorkspaceFailureOutputPane.cs]: curlies

**Comment by jasonmalinowski** [comment]
> Is focus only set to the IntelliSense Errors window the first time a message is written to it (i.e. it never occurs again unless the IDE is restarted)? If so, the current behavior sounds like what we would want.

It does, but because of the workflow TemporaryGeneratedFiles, MSBuild-generated TFM files, etc., that the odds of you getting an error is quite high. And those errors are 100% useless.

**Comment by jasonmalinowski** [comment]
Frankly, @Pilchie, we added this pane because it's what the old IDE did. Do we have _any_ evidence this has helped _anyone_ diagnose _anything_? :smile:

**Comment by Pilchie** [comment]
I'm not aware of any such evidence :)

**Comment by jasonmalinowski** [comment]
Well, I vote for just deleting the thing. Somebody to second the motion?

**Comment by davkean** [comment]
@dotnet/roslyn-ide Ran into this again. Give no one has taken up to delete the Output window, can we please look over this and get code review sign-off? This can be a stop gap until someone deletes the category.

**Comment by davkean** [comment]
This is no longer a work in progress; I've successfully tested this and confirmed it works. I will merge into tomorrow if there is no more feedback.

**Comment by davkean** [comment]
@jinujoseph Can I get approval for this?

File: src/VisualStudio/Core/Def/Implementation/Workspace/WorkspaceFailureOutputPane.cs [hunk 2]
```diff
@@ -49,26 +49,58 @@ private IVsOutputWindowPane OutputPane
...
                     }
                 }
 
                 return _doNotAccessDirectlyOutputPane;
             }
         }
+
+        private IVsOutputWindowPane CreateOutputPane(IVsOutputWindow outputWindow)
+        {
+            // Try to get the workspace pane if it has already been registered
```
Comment (sharwell) [inline src/VisualStudio/Core/Def/Implementation/Workspace/WorkspaceFailureOutputPane.cs]: ❕ Should assert UI thread affinity (`AssertIsForeground()`)

File: src/VisualStudio/Core/Def/Implementation/Workspace/WorkspaceFailureOutputPane.cs [hunk 3]
```diff
@@ -49,26 +49,58 @@ private IVsOutputWindowPane OutputPane
...
+            {
+                return pane;
+            }
+
+            return null;
+        }
+
+        private static Guid GetActivePane(IVsOutputWindow outputWindow)
+        {
+            if (outputWindow is IVsOutputWindow2 outputWindow2)
```
Comment (sharwell) [inline src/VisualStudio/Core/Def/Implementation/Workspace/WorkspaceFailureOutputPane.cs]: ❕ Should assert UI thread affinity (`AssertIsForeground()`)

File: src/VisualStudio/Core/Def/Implementation/Workspace/WorkspaceFailureOutputPane.cs [hunk 4]
```diff
@@ -49,26 +49,58 @@ private IVsOutputWindowPane OutputPane
...
+                    return activePaneGuid;
+                }
+            }
+
+            return Guid.Empty;
+        }
+
+        public static void ActivatePane(IVsOutputWindow outputWindow, Guid paneGuid)
+        {
+            if (ErrorHandler.Succeeded(outputWindow.GetPane(ref paneGuid, out var pane)))
```
Comment (sharwell) [inline src/VisualStudio/Core/Def/Implementation/Workspace/WorkspaceFailureOutputPane.cs]: ❕ Should assert UI thread affinity (`AssertIsForeground()`)

File: src/VisualStudio/Core/Def/Implementation/Workspace/WorkspaceFailureOutputPane.cs [hunk 5]
```diff
@@ -49,26 +49,58 @@ private IVsOutputWindowPane OutputPane
...
+            {
+                pane.Activate();
```
Comment (sharwell) [inline src/VisualStudio/Core/Def/Implementation/Workspace/WorkspaceFailureOutputPane.cs]: ❔ Does this set focus to the Output Window, or just change the selection? (possible question for @olegtk)

**Comment by sharwell** [comment]
I am neutral as to whether or not we accept this change. Functionally, I believe it achieves the intended purpose and the code behavior is clear (as can be) aside from the one question I added.

#### The following items prevent me from landing on the side of full support

1. The messages described #12202 and #27964 are not the only messages capable of using this window. A review of code paths leading to this window suggest other more serious situations could be revealed by this window, and hiding the window would cause a user to miss these messages.
2. During an initial attempt to remove the specific messages mentioned in the previous issues (https://github.com/dotnet/roslyn/pull/26230#issuecomment-382471716), I became concerned that the underlying cause of these messages is a bug in Roslyn's handling of file change notifications and/or incremental updates. The IntelliSense engine should be accounting for the possibility of asynchronous modifications to the file, and not strictly assuming they will exist. Even more concerning to me is sensitivity to asynchronous file modifications likely means Roslyn is capable of producing `Compilation` instances that represent torn state - i.e. the immutable snapshot compilation describes a state which never actually existed.

In short, this PR adds a bunch of code to hide what is likely a more serious bug in the project.

#### The following items prevent me from landing on the side of blocking

1. I agree that the messages described in the bugs this PR works around are annoying and generally not helpful
2. We have not established a timeline for fixing the bugs mentioned above, so without a workaround things are likely to stay as they are

#### Summary

I defer to the rest of the IDE team to decide whether or not to take this change in light of the information above. If they decide to take the change, I believe my code review is relevant.

File: src/VisualStudio/Core/Def/Implementation/Workspace/WorkspaceFailureOutputPane.cs (hunk 5 above)
Comment (olegtk) [reply src/VisualStudio/Core/Def/Implementation/Workspace/WorkspaceFailureOutputPane.cs]: this eventually calls IVsWindowFrame.Show(), which makes Output window visible and sets focus to its content.

Comment (davkean) [reply src/VisualStudio/Core/Def/Implementation/Workspace/WorkspaceFailureOutputPane.cs]: This does not activate the Output window, see: https://devdiv.visualstudio.com/DevDiv/_git/VS?path=%2Fsrc%2Fenv%2Fmsenv%2Fcore%2Foutpwin.cpp&version=GBlab%2Fvsumain&line=2979&lineStyle=plain&lineEnd=2980&lineStartColumn=1&lineEndColumn=1. Notice how it passes `FALSE` to `setFocus`.

File: src/VisualStudio/Core/Def/Implementation/Workspace/WorkspaceFailureOutputPane.cs (hunk 2 above)
Comment (davkean) [reply src/VisualStudio/Core/Def/Implementation/Workspace/WorkspaceFailureOutputPane.cs]: Fixed.

File: src/VisualStudio/Core/Def/Implementation/Workspace/WorkspaceFailureOutputPane.cs (hunk 3 above)
Comment (davkean) [reply src/VisualStudio/Core/Def/Implementation/Workspace/WorkspaceFailureOutputPane.cs]: Fixed.

File: src/VisualStudio/Core/Def/Implementation/Workspace/WorkspaceFailureOutputPane.cs (hunk 4 above)
Comment (davkean) [reply src/VisualStudio/Core/Def/Implementation/Workspace/WorkspaceFailureOutputPane.cs]: Fixed.

**Comment by davkean** [comment]
Sam, none of the above concerns should block this check-in, we are not hiding the pane. We just don't make it steal focus the very _first_ time it is created. It never steals focus after that.

**Comment by jasonmalinowski** [comment]
Yeah, totally OK with this change. Regarding these messages:

1. The nature of the file system means these issues can be legitimate problems, or entirely transient issues that we will handle 100% correctly; the messages give no way to determine which of those two it is.
2. I don't think I can recall of a single instance where these were ever used to diagnose any problem.
3. I can think of plenty of times people saw the errors and panicked and got confused by the problem.

I'd still vote just delete, but admittedly haven't found the time myself to make a change now. And no reason to hold this up for the undetermined future.

**Comment by CyrusNajmabadi** [comment]
> The nature of the file system means these issues can be legitimate problems, or entirely transient issues that we will handle 100% correctly; the messages give no way to determine which of those two it is.

One option to consider:

During build collect the information, but do not display it until the end.  At the end, if we thought we couldn't read a file *but now we can*, suppress do not report that problem.

--

> I'd still vote just delete, but admittedly haven't found the time myself to make a change now. And no reason to hold this up for the undetermined future.

I agree with this.  It just doesn't actually have any positive benefits to the user.  And it def have negative ones.

**Comment by jasonmalinowski** [comment]
> During build collect the information, but do not display it until the end. At the end, if we thought we couldn't read a file but now we can, suppress do not report that problem.

The information isn't collected during build; the simple example of these is you open a solution that pulls in an autogenerated file, but that file isn't generated until the first real build. We'll give this this error, but we have no way to know whether this is a project file being broken, or just something that will be fixed by a build perhaps minutes or hours later.

**Comment by CyrusNajmabadi** [comment]
> The information isn't collected during build; the simple example of these is you open a solution that pulls in an autogenerated file, but that file isn't generated until the first real build. We'll give this this error, but we have no way to know whether this is a project file being broken, or just something that will be fixed by a build perhaps minutes or hours later.

That's fair.  however, for the bug i reported (https://github.com/dotnet/roslyn/issues/27964), the issue is files presumably getting deleted/regenerated during build.

So at least some of the issues could potentially be addressed, if not all of them.

**Comment by jasonmalinowski** [comment]
@jinujoseph This is ready for M2 approval then. @davkean once that's approved we can merge; you want the honors or should I?

**Comment by davkean** [comment]
This error occurs simply by deleting a file in CPS projects. No auto-generation needed.

**Comment by jinujoseph** [comment]
Approved to merge for 15.8.Preview4
